### PR TITLE
virtualbox: 5.2.28 -> 6.0.6

### PR DESCRIPTION
**Pull request by ptrhlm**
###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

###### Things to do
  - ~Build with hardening doesn't start~
  - Tests

### CODE REVIEW

**Comment by jtojnar** [comment]
Changes: https://www.virtualbox.org/wiki/Changelog-6.0#v0

**Comment by bendlas** [comment]
I've tested this PR on my machine, after upgrading to `6.0.4` https://github.com/bendlas/nixpkgs/commit/1a73f8b307f5db56bd4cd753454f32337034e2db

It doesn't seem to find its extension pack (installed in libexec due to a non-hardened build)

**Comment by ptrhlm** [comment]
I removed the WIP status, but I couldn't get the tests to work.
Most likely it's because nested virtualization stopped working (tests run VirtualBox guest inside of QEMU host). 

Everything else should work, including 3D accel.

File: nixos/modules/virtualisation/virtualbox-host.nix [hunk 1]
```diff
@@ -102,7 +102,8 @@ in
       "VBoxNetNAT"
       "VBoxSDL"
       "VBoxVolInfo"
-      "VirtualBox"
+      "VirtualBoxVM"
+      "VBoxHeadless"
```
Comment (wkral) [inline nixos/modules/virtualisation/virtualbox-host.nix]: This appears to be a repetition of the first line in this list.
```suggestion
```

Comment (ptrhlm) [reply nixos/modules/virtualisation/virtualbox-host.nix]: Thanks for pointing out. Fixed

**Comment by Laaas** [comment]
What's blocking this?

**Comment by kalbasit** [comment]
I tried it, it works on my side. @Infinisil I'd like to merge this if it's alright with you.

**Comment by flokli** [comment]
I see three open questions:

 - hardening
 - extension pack
 - nixos tests

I'm not entirely against merging this while these don't work yet (it doesn't currently build with linuxPackages_latest at all), but we should still pick this up :-)

**Comment by wkral** [comment]
Please merge #55596 before or with this merge as that PR implies that this would break vagrant since the current version in unstable doesn't support virtualbox 6.

**Comment by kalbasit** [comment]
@wkral thank you for letting me know, I will merge #55596 after this one.

@flokli Let's hear from @Infinisil first in this case!

**Comment by ptrhlm** [comment]
@flokli As far I can tell, hardening and extension pack work. 

I wasn't able to get the NixOS tests to run and fixing them is beyond my abilities.

**Comment by flokli** [comment]
@aszlig do you have time to dig into the virtualbox tests?

**Comment by bendlas** [comment]
> As far I can tell, hardening and extension pack work.

Have you tested adding USB 2/3 and other advanced hardware? That failed for me i.e. I couldn't open any of my saved machines w/o deactivating all the hardware. Said it couldn't find the extension pack.

If you think this is working, I'll be happy to give it another shot.

**Comment by artemist** [comment]
I also had a problem when recovering saves (it talked about pgm). It may have been a change in the state save format, since when I discarded and restarted the machine everything was working, including NVMe and USB 2.0. The extpack is also listed as functioning and installed.

```
$ VBoxManage list extpacks
Extension Packs: 1
Pack no. 0:   Oracle VM VirtualBox Extension Pack
Version:      6.0.4
Revision:     128413
Edition:      
Description:  USB 2.0 and USB 3.0 Host Controller, Host Webcam, VirtualBox RDP, PXE ROM, Disk Encryption, NVMe.
VRDE Module:  VBoxVRDP
Usable:       true 
Why unusable: 
```

If we have different problems, what do you get when you list extension packs?

**Comment by jb55** [comment]
So I gave this a spin. It's pretty much unusable for me, I had to rollback.

Issues:

  - Video was flickering, everything is much slower
  - Guest additions doesn't install, gets an network error
  - Couldn't get oracle vm extension to work for usb 2.0 support

**Comment by flokli** [comment]
virtualbox also discontinued supporting `i686-linux` with version 6, and claim they'll still support version 5.2 (funnily inside not inside their release notes, just on the [website](https://www.virtualbox.org/wiki/Downloads))

> Please also use version 5.2 if you still need support for 32-bit hosts, as this has been discontinued in 6.0. Version 5.2 will remain supported until July 2020.

**Comment by ptrhlm** [comment]
> So we should drop i686-linux from `meta`.

@flokli I removed "i686-linux" from `meta`. 
Maybe it is worth considering adding a separate package with VirtualBox version 5.2 for this platform.

> Have you tested adding USB 2/3 and other advanced hardware?

@bendlas Yes, I tried with USB2/3. 
I also looked at your nixos configuration.  You try to enable extension pack with `virtualbox.enableExtensionPack = true;` . It never worked for me this way, so I enabled it using virtualbox host module:

```nix
{
  virtualisation = {
    virtualbox.host = {
      enable = true;
      enableExtensionPack = true;
    };
  };
}
```

I also just realized that I disabled hardening in my configuration. I will try to rebuild and fix any problems.

**Comment by flokli** [comment]
I packaged virtualbox 5.2.26 in https://github.com/NixOS/nixpkgs/pull/56210, which should remove the 19.03 blocker at least.

**Comment by artemist** [comment]
Considering how VirtualBox 6 drops support for i686, and NixOS still supports i686, we may want to make virtualbox versions 5 and 6 different packages, then just use `virtualization.virtualbox.host.package` to decide between them

**Comment by flokli** [comment]
@artemist in fact, I was planning on abandoning `cfg.package`, and instead just have switches for hardening and extension pack, easing some parts of the wrapper to solve https://github.com/NixOS/nixpkgs/issues/34796#issuecomment-365070156.

Unfortunately, I never found the time to polish that up, and probably won't get to that until 19.03 branch-off. I'm also not sure if we can fix virtualbox 6 until then either.

Maybe better stick with virtualbox 5 for 19.03? On the other hand, how much do people rely on virtualbox on i686?

**Comment by kalbasit** [comment]
I merged down #56210 and have removed the blocker label from this pull request. We can work on getting this out after 19.03.

**Comment by bendlas** [comment]
> I also looked at your nixos configuration. You try to enable extension pack with `virtualbox.enableExtensionPack = true;` . It never worked for me this way, so I enabled it using virtualbox host module:

ok, this is embarrassing. with the system config, everything seems to work now. (the menu-based download for the guest additions fails with repeatably with unspecified network errors, probably not something we should fix)

sorry for not a/b testing and thanks for debugging my configuration @ptrhlm ;-)

as far as I'm concerned, the Extension Pack issue is resolved.

**Comment by jb55** [comment]
A few updates since I last tried this:

William Casarin <notifications@github.com> writes:
> So I gave this a spin. It's pretty much unusable for me, I had to rollback.
>
> Issues:
>
>   - Video was flickering, everything is much slower

The flickering was happened with 3d/2d acceleration turned on. When I
turn it off, no more ficker

>   - Couldn't get oracle vm extension to work

I didn't have the nix option set, this works now.

**Comment by flokli** [comment]
@ptrhlm this doesn't apply anymore, can you rebase?

Is there anybody who can dedicate some time to debug the virtualbox test breakage?

**Comment by bendlas** [comment]
> Is there anybody who can dedicate some time to debug the virtualbox test breakage?

FWIW, here is a log of the nixos tests failing: https://gist.githubusercontent.com/bendlas/02111b84e4cce6916988047f5fc9ff12/raw/9d64e28bc6106d70a2f7795b9c667babf6c1415c/vbox-tests.log

**Comment by devhell** [comment]
So, what's happening with this?

**Comment by flokli** [comment]
IMHO, we can't bump virtualbox to 6.x if we can't get the tests to work.

There's too much that can break randomly and to test manually.

I'm not sure what exactly was dropped in virtualbox 6.0 - docs mention 32bit support (which is why they keep 5.2 around until 2020), but maybe this broke nested virtualization too.

@devhell if you can devote some time to debug this, it'd be much appreciated :-)

**Comment by devhell** [comment]
@flokli Thanks for the info. As so many of us, I wish I had the time, and in this case also the skills required to work out the tests. :(

**Comment by tonylotts** [comment]
I'm happy to test this on macOS, although I'm a net new to Nix.  Can someone guide me on the appropriate process to check this out, and test it?

**Comment by tonylotts** [comment]
I've created a Darwin specific package, here https://github.com/tonylotts/nixpkgs/commit/d0d87fcb4d5d7ed81f5c32d5a338c8322a662c22

**Comment by flokli** [comment]
@tonylotts I don't think virtualbox 6.0.4 completly broken for linux, it's just that the nixos vm tests don't work anymore (maybe due to the nested virtualization happening here) so we don't spot things suddenly breaking, which is why I'm hesitant to bump it without fixing the tests too…

Note the darwin-specific `virtualbox6` derivation you proposed doesn't build VirtualBox, but simply unpacks a `.dmg`.

**Comment by tonylotts** [comment]
@flokli, yes I've purposely used the pre-built binary; although I acknowledge that the community preference is to build from source.  As I understand it, pre-built binaries are cached.  When one   In the case that VirtualBox is already available in pre-built form, can/should that be cached?

**Comment by flokli** [comment]
Even when there's a from-source build in nixpkgs, it should be built by Hydra and available from cache.nixos.org.

This currently isn't working for virtual box with extension pack enabled, as IIRC its license forbids redistribution of the extpack itself, but for the non-extpack variant it should be in the cache.

**Comment by tonylotts** [comment]
Got it.

**Comment by tonylotts** [comment]
It turns out that the macOS version of VirtualBox provides kernel extensions.  A macOS derivation for VirtualBox would require nix-darwin, and support for kernel extensions https://github.com/LnL7/nix-darwin/pull/91.

**Comment by flokli** [comment]
@tonylotts so no matter what variant we use, this depends on https://github.com/LnL7/nix-darwin/pull/91 or a successor to be merged?

**Comment by devhell** [comment]
6.0.6 has been released. Not that it matters that much, I guess, right now.

**Comment by tonylotts** [comment]
@flokli, yes

**Comment by flokli** [comment]
@devhell I tried bumping to 6.0.6, but it seems some patches need to be adjusted. Did you do that already somewhere?

**Comment by flokli** [comment]
The release notes mention some **nested (AMD) virtualization fixes** - I guess we still should see if things have improved.

**Comment by devhell** [comment]
@flokli I've not been able to, no. Sorry. :/

**Comment by flokli** [comment]
@ptrhlm could you get to that?

**Comment by ptrhlm** [comment]
@flokli I've removed unnecessary kernel module patches. Also, I moved patching of kernel module sources to main package, so now unaligned patch will fail earlier.

Tests still fail on my Intel machine.

**Comment by flokli** [comment]
@ptrhlm thanks for the cleanup! I gave the tests a try on an AMD machine too, unfortunately, they fail there, too:

`nix log /nix/store/pnbnmxz36kxdd33b8mj9v6qh2i7hmiws-vm-test-run-virtualbox-headless.drv` looks most scary:
```
machine# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
machine# !!
machine# !!         VCPU0: Guru Meditation -2403 (VERR_TRPM_DONT_PANIC)
machine# !!
machine# !! TRAP=0e ERRCD=0000000000000010 CR2=00000000a082bf60 EIP=a082bf60 Type=0 cbInstr=ff
machine# !! EIP is not in any code known to VMM!
machine# !!
machine# !!
machine# !!
```

**Pull request closed without merging**